src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,454 @@
+
+3. Once the isolate is spawned, the main isolate waits for the result.
+   Because the `ReceivePort` class implements `Stream`,
+   the [`first`][] property is an easy way to get
+   the single message that the worker isolate sends.
+
+[`first`]: {{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-async/Stream/first.html
+
+The spawned isolate executes the following code:
+
+```dart
+Future _readAndParseJson(SendPort p) async {
+  final fileData = await File(filename).readAsString();
+  final jsonData = jsonDecode(fileData);
+  Isolate.exit(p, jsonData);
+}
+```
+
+The relevant statement is the last one, which exits the isolate,
+sending `jsonData` to the passed-in `SendPort`.
+Message passing between isolates normally involves data copying,
+and thus can be slow and increases with the size of the message —
+O(n) in [big O notation][].
+However, when you send the data using `Isolate.exit()`,
+then the memory that holds the message in the exiting isolate isn’t copied,
+but instead is transferred to the receiving isolate.
+That transfer is quick and completes in constant time — O(1).
+
+[big O notation]: https://en.wikipedia.org/wiki/Big_O_notation
+
+{{site.alert.version-note}}
+  `Isolate.exit()` was added in 2.15.
+  Previous releases support only explicit message passing,
+  using `Isolate.send()` as shown in the next section's example.
+{{site.alert.end}}
+
+The following figure illustrates the communication between
+the main isolate and the worker isolate:
+
+![A figure showing the previous snippets of code running in the main isolate and in the worker isolate](/guides/language/concurrency/images/isolate-api.png)
+
+
+### Sending multiple messages between isolates
+
+If you need more communication between isolates,
+then you need to use the [`send()` method][] of `SendPort`.
+One common pattern, which the following figure shows,
+is for the main isolate to send a request message to the worker isolate,
+which then sends one or more reply messages.
+
+![A figure showing the main isolate spawning the isolate and then sending a request message, which the worker isolate responds to with a reply message; two request-reply cycles are shown](/guides/language/concurrency/images/isolate-custom-bg-worker.png)
+
+For examples of sending multiple messages,
+see the following [isolate samples][]:
+
+* [send_and_receive.dart][],
+  which shows how to send a message from
+  the main isolate to the spawned isolate.
+  It’s otherwise similar to the preceding example.
+* [long_running_isolate.dart][],
+  which shows how to spawn a long-running isolate that
+  receives and sends multiple times.
+
+{% assign samples = "https://github.com/dart-lang/samples/tree/master/isolates" %}
+
+[isolate samples]: {{ samples }}
+[send_and_receive.dart]: {{ samples }}/bin/send_and_receive.dart
+[long_running_isolate.dart]: {{ samples }}/bin/long_running_isolate.dart
+
+{{site.alert.tip}}
+  The `send()` method can be slow.
```

Comment:
It's one of the main APIs that's used to communicate between isolates (it also got much faster recently). The performance of `Isolate.exit()` and `SendPort.send()` isn't that much different actually.

=> I would really not discourage users by saying `can be slow. avoid when possible`.

Instead I'd focus on the actual differences between the two:

Using `SendPort.send()` will transitively copy the message being sent (except for immutable objects such as strings which are shared). The receiver will receive this copy. To avoid such a memory copy one can use `Isolate.exit()` which will hand the existing object graph to the receiver without copying it, while exiting the current isolate.

Reply:
I learned earlier today that Isolate.send is fast for isolates in a group.

Should we perhaps remove the "`send()` method can be slow." part from here, and then in the section below for Isolate.spawnUri add something like: "Also note that Isolate.send() is slower between isolates not in an isolate group" ?

Reply:
Wouldn't sending any message between isolate groups be slower? (No need to single out send(), is there?)

Reply:
Of course, Isolate.exit() isn't supported between different isolate groups, so it's kind of equivalent.

Reply:
Sending `TransferableTypedData` will be similarly fast regardless of whether isolates are in the same group or in different groups.